packages/expo-dev-menu/android/src/main/java/expo/modules/devmenu/DevMenuManager.kt

```diff
@@ -0,0 +1,227 @@
+        reactContext.addLifecycleEventListener(this)
+      }
+    }
+  }
+
+  //endregion
+
+  //region shake detector
+
+  /**
+   * Starts [ShakeDetector] if it's not running yet.
+   */
+  private fun maybeStartDetectingShakes(context: Context) {
+    if (shakeDetector != null) {
+      return
+    }
+    shakeDetector = ShakeDetector { onShakeGesture() }.apply {
+      start(context.getSystemService(Context.SENSOR_SERVICE) as SensorManager)
+    }
+  }
+
+  /**
+   * Handles shake gesture which simply toggles the dev menu.
+   */
+  private fun onShakeGesture() {
+    if (settings?.motionGestureEnabled == true) {
+      delegateActivity?.let {
+        toggleMenu(it)
+      }
+    }
+  }
+
+  //endregion
+
+  //region delegate's LifecycleEventListener
+
+  override fun onHostResume() {
+    delegateReactContext?.removeLifecycleEventListener(this)
+
+    if (shouldLaunchDevMenuOnStart) {
+      shouldLaunchDevMenuOnStart = false
+      delegateActivity?.let {
+        openMenu(it)
+      }
+    }
+  }
+
+  override fun onHostPause() = Unit
+
+  override fun onHostDestroy() = Unit
+
+  //endregion
+
+  //region DevMenuManagerProtocol
+
+  override fun openMenu(activity: Activity) {
+    session = DevMenuSession(delegate!!.reactInstanceManager(), delegate!!.appInfo())
+    activity.startActivity(Intent(activity, DevMenuActivity::class.java))
+  }
+
+  override fun closeMenu() {
+    // Sends an event to JS triggering the animation that collapses the dev menu.
+    hostReactContext
+      ?.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
+      ?.emit("closeDevMenu", null)
+  }
+
+  override fun hideMenu() {
+    hostActivity?.finish()
+  }
+
+  override fun toggleMenu(activity: Activity) {
+    if (hostActivity?.isDestroyed == false) {
+      closeMenu()
+    } else {
+      openMenu(activity)
+    }
+  }
+
+  override fun onKeyEvent(keyCode: Int, event: KeyEvent): Boolean {
+    if (settings?.keyCommandsEnabled != true) {
+      return false
+    }
+
+    val keyCommand = KeyCommand(keyCode, event.modifiers)
+    delegateActions.forEach {
+      if (it.keyCommand == keyCommand) {
+        it.action()
+        return true
+      }
+    }
+
+    return false
+  }
+
+  override fun setDelegate(newDelegate: DevMenuDelegateProtocol) {
+    // remove event listener for old delegate
+    delegateReactContext?.removeLifecycleEventListener(this)
+
+    maybeStartDetectingShakes(devMenuHost.getContext())
+    delegate = newDelegate.run {
+      if (reactInstanceManager().currentReactContext == null) {
+        reactInstanceManager().addReactInstanceEventListener(this@DevMenuManager::handleLoadedDelegate)
+      } else {
+        handleLoadedDelegate(reactInstanceManager().currentReactContext!!)
+      }
+      this
+    }
+  }
+
+  override fun dispatchAction(actionId: String) {
+    delegateMenuItems.forEach {
+      if (it is DevMenuAction && it.actionId == actionId) {
+        it.action()
+        return
+      }
+    }
```

Comment:
```suggestion
    delegateMenuItems
      .find({ it -> it is DevMenuAction && it.actionId == actionId })
      ?.let { ::action }
```
Isn't it easier to follow?
I'm not perfectly sure what `return` refers to 🤔  `forEach` scope or `function` scope?
I'm not perfectly sure about syntax as well 👅

Reply:
`return` with no label always refers to nearest `fun` scope 😉

https://kotlinlang.org/docs/reference/returns.html